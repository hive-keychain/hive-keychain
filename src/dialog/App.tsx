import { DialogCommand } from '@reference-data/dialog-message-key.enum';
import React, { useEffect, useState } from 'react';
import DialogError from 'src/dialog/pages/error';
import Register from 'src/dialog/pages/register';
import RequestConfirmation from 'src/dialog/pages/request-confirmation';
import RequestResponse from 'src/dialog/pages/request-response';
import SignTransaction from 'src/dialog/pages/sign-transaction';
import Unlock from 'src/dialog/pages/unlock';
import BrowserUtils from 'src/utils/browser.utils';
import './../analytics/analytics/gtag';
import './dialog.scss';

const App = () => {
  const [data, setData] = useState<any>({});
  const initGoogleAnalytics = () => {
    window.dataLayer = window.dataLayer || [];
    window.gtag = function gtag() {
      window.dataLayer.push(arguments); // eslint-disable-line
    };
    window.gtag('js', new Date());
    window.gtag(
      'config',
      process.env.GOOGLE_ANALYTICS_TAG_ID || 'G-1LRCTFLVBH',
      {
        page_path: '/popup',
      },
    );
    window.gtag('set', 'checkProtocolTask', () => {}); // Disables file protocol checking.

    window.gtag('event', 'navigation', {
      page: 'dialog',
    });
  };

  useEffect(() => {
    initGoogleAnalytics();
    chrome.runtime.onMessage.addListener(async function (
      data,
      sender,
      sendResp,
    ) {
      if (data.command === DialogCommand.READY) {
        return BrowserUtils.sendResponse(true, sendResp);
      } else if (Object.values(DialogCommand).includes(data.command)) {
        setData(data);
      }
      chrome.windows.update((await chrome.windows.getCurrent()).id!, {
        focused: true,
      });
    });
  }, []);

  const renderDialogContent = (data: any) => {
    switch (data.command) {
      case DialogCommand.UNLOCK:
        return <Unlock data={data} />;
      case DialogCommand.WRONG_MK:
        return <Unlock data={data} wrongMk index={Math.random()} />;
      case DialogCommand.SEND_DIALOG_ERROR:
        return <DialogError data={data} />;
      case DialogCommand.REGISTER:
        return <Register data={data} />;
      case DialogCommand.SEND_DIALOG_CONFIRM:
        return <RequestConfirmation data={data} />;
      case DialogCommand.ANSWER_REQUEST:
        return <RequestResponse data={data} />;
      case DialogCommand.SIGN_WITH_LEDGER:
        return <SignTransaction data={data} />;
      default:
        return null;
    }
  };

  return <div className="dialog">{renderDialogContent(data)}</div>;
};

export default App;
