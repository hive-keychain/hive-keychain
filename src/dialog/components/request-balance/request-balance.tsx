import { Rpc } from '@interfaces/rpc.interface';
import React, { useEffect, useState } from 'react';
import RequestItem from 'src/dialog/components/request-item/request-item';
import AccountUtils from 'src/utils/account.utils';
import CurrencyUtils, { BaseCurrencies } from 'src/utils/currency.utils';

type Props = {
  amount: number;
  currency: string;
  username?: string;
  rpc: Rpc;
};

const RequestBalance = ({ rpc, username, amount, currency }: Props) => {
  const [balance, setBalance] = useState('');
  const [newBalance, setNewBalance] = useState('');
  const cur = currency.toLowerCase();
  useEffect(() => {
    if (username) {
      init(username);
    }
  }, [username]);

  const init = async (username: string) => {
    const account = await AccountUtils.getExtendedAccount(username);
    const currencyParsed = CurrencyUtils.getCurrencyLabel(
      currency,
      rpc.testnet,
    );
    const currentBalance = (
      (cur === BaseCurrencies.HIVE
        ? account.balance
        : account.hbd_balance) as string
    ).split(' ')[0];
    const newBalance = (parseFloat(currentBalance) - amount).toFixed(3);
    setBalance(`${currentBalance} ${currencyParsed}`);
    setNewBalance(`${newBalance} ${currencyParsed}`);
  };

  return (
    <RequestItem
      title="dialog_balance"
      content={
        balance.length
          ? parseFloat(newBalance) < 0
            ? chrome.i18n.getMessage('dialog_insufficient_balance')
            : `${balance} => ${newBalance}`
          : '...'
      }
    />
  );
};

export default RequestBalance;
