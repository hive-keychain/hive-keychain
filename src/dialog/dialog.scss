@import '/src/popup/variable.scss';

@font-face {
  font-family: Futura;
  src: url('/assets/fonts/Futura-Boo.otf');
  font-weight: 400;
}

html {
  height: 100%;
}
body {
  margin: 0px;
  height: 100%;
  overflow: hidden;
  #root {
    display: block;
    height: 100%;
  }
}
//TODO: Same as import / would be better of as variable
.dialog {
  font-family: Futura !important;
  height: 100%;
  background: url('/assets/images/hive-watermark.png') 0% 0% no-repeat
      padding-box,
    linear-gradient(330deg, #59626c 0%, #000000 100%) 0% 0% no-repeat
      padding-box !important;
  color: white;
  padding: 32px;
  box-sizing: border-box;
  display: flex;
  flex-direction: column;
  font-size: 16px;
}

pre {
  font-family: Futura !important;
  margin: 0;
}

::-webkit-scrollbar-track {
  -webkit-box-shadow: unset;
  box-shadow: unset;
  border-radius: 4px;
  background-color: transparent;
}

::-webkit-scrollbar {
  width: 6px;
  background-color: transparent;
}

::-webkit-scrollbar-thumb {
  border-radius: 4px;
  -webkit-box-shadow: inset 0 0 6px rgba(0, 0, 0, 0.3);
  box-shadow: inset 0 0 6px rgba(0, 0, 0, 0.3);
  background-color: #e31337;
}
