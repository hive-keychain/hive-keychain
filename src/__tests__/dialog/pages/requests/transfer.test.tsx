import { BackgroundCommand } from '@reference-data/background-message-key.enum';
import { render, waitFor } from '@testing-library/react';
import React from 'react';
import Transfer from 'src/dialog/pages/requests/transfer';
import appMocks from 'src/__tests__/dialog/mocks/app-mocks';
import alButton from 'src/__tests__/utils-for-testing/aria-labels/al-button';
import { Tests_Client } from 'src/__tests__/utils-for-testing/classes/dialog/request-balance';
import phishing from 'src/__tests__/utils-for-testing/data/phishing';
import requestTransfer from 'src/__tests__/utils-for-testing/data/props/dialog/request-transfer';
import objects from 'src/__tests__/utils-for-testing/helpers/objects';
import mockPreset from 'src/__tests__/utils-for-testing/preset/mock-preset';
import { clickAwait } from 'src/__tests__/utils-for-testing/setups/events';
import { PropsRequestTransfer } from 'src/__tests__/utils-for-testing/types/props-types';
describe('transfer tests:\n', () => {
  const { methods, spies, mocks } = appMocks;
  const { props } = requestTransfer;
  methods.config();
  beforeEach(() => {
    mockPreset.setOrDefault({});
    mocks.dHiveHiveIO(Tests_Client);
  });
  it('Must show Transfer dialog(Anonymous, encrypted memo, phishing warning)', async () => {
    const clonedProps = objects.clone(props) as PropsRequestTransfer;
    clonedProps.data.to = phishing.accounts.data[0];
    clonedProps.data.memo = '#To Encrypt memo';
    clonedProps.data.amount = '1.000';
    const { asFragment } = render(<Transfer {...clonedProps} />);
    await waitFor(() => {});
    expect(asFragment()).toMatchSnapshot(
      'Request Transfer Anonymous, encrypted memo, phishing warning',
    );
  });
  it('Must show Transfer dialog(regular memo)', async () => {
    const clonedProps = objects.clone(props) as PropsRequestTransfer;
    clonedProps.data.memo = 'Regular memo!';
    clonedProps.data.amount = '1.000';
    const { asFragment } = render(<Transfer {...clonedProps} />);
    await waitFor(() => {});
    expect(asFragment()).toMatchSnapshot('Request Transfer regular memo');
  });
  it('Must show Transfer dialog(empty memo)', async () => {
    const clonedProps = objects.clone(props) as PropsRequestTransfer;
    clonedProps.data.memo = '';
    clonedProps.data.amount = '1.000';
    const { asFragment } = render(<Transfer {...clonedProps} />);
    await waitFor(() => {});
    expect(asFragment()).toMatchSnapshot('Request Transfer empty memo');
  });
  it('Must call window.close when clicking on cancel', async () => {
    const clonedProps = objects.clone(props) as PropsRequestTransfer;
    clonedProps.data.amount = '1.000';
    render(<Transfer {...clonedProps} />);
    await clickAwait([alButton.dialog.cancel]);
    expect(spies.closeWindow).toBeCalledTimes(1);
  });
  it('Must call onConfirm when clicking on confirm', async () => {
    const clonedProps = objects.clone(props) as PropsRequestTransfer;
    clonedProps.data.amount = '1.000';
    render(<Transfer {...{ ...clonedProps, onConfirm: spies.onConfirm }} />);
    await clickAwait([alButton.dialog.confirm]);
    expect(spies.onConfirm).toBeCalledTimes(1);
  });
  it('Must call sendMessage when clicking on confirm', async () => {
    const clonedProps = objects.clone(props) as PropsRequestTransfer;
    clonedProps.data.amount = '1.000';
    const spySendMessage = jest.spyOn(chrome.runtime, 'sendMessage');
    render(<Transfer {...clonedProps} />);
    await clickAwait([alButton.dialog.confirm]);
    expect(spySendMessage).toBeCalledWith({
      command: BackgroundCommand.ACCEPT_TRANSACTION,
      value: {
        data: clonedProps.data,
        tab: clonedProps.tab,
        domain: clonedProps.domain,
        keep: false,
      },
    });
  });
});
