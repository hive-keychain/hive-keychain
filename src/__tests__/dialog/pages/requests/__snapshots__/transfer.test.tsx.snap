// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`transfer tests:
 Must show Transfer dialog(Anonymous, encrypted memo, phishing warning): Request Transfer Anonymous, encrypted memo, phishing warning 1`] = `
<DocumentFragment>
  <div
    aria-label="dialog-operation-component"
    class="operation"
  >
    <div>
      <h2
        aria-label="dialog-header-operation"
        class="dialog-header"
      >
        Transfer
      </h2>
      <div
        aria-label="operation_header"
        class="operation_header operation-red"
      >
        @aalpha has been listed as a potential phishing account, check the spelling and make sure it is the account you want to transfer to.
      </div>
    </div>
    <div
      class="operation_body"
    >
      <h3>
        Account
      </h3>
      <div
        aria-label="operation_item_content"
        class="operation_item_content "
      >
        @keychain.tests
      </div>
      <h3>
        To
      </h3>
      <div
        aria-label="operation_item_content"
        class="operation_item_content "
      >
        @aalpha
      </div>
      <h3>
        Amount
      </h3>
      <div
        aria-label="operation_item_content"
        class="operation_item_content "
      >
        1.000 HIVE
      </div>
      <h3>
        Balance
      </h3>
      <div
        aria-label="operation_item_content"
        class="operation_item_content "
      >
        100 HIVE =&gt; 99.000 HIVE
      </div>
      <h3>
        Memo
      </h3>
      <div
        aria-label="operation_item_content"
        class="operation_item_content "
      >
        #To Encrypt memo (encrypted)
      </div>
    </div>
    <div
      class="operation_footer no-whitelist"
    >
      <div
        aria-label="whitelist-operation"
        class="whitelist_operation"
      />
      <div
        class="operation_buttons "
      >
        <button
          aria-label="dialog_cancel-button"
          class="submit-button stroked "
        >
          <div
            class="button-label"
          >
            Cancel 
          </div>
        </button>
        <button
          aria-label="dialog_confirm-button"
          class="submit-button raised "
        >
          <div
            class="button-label"
          >
            Confirm 
          </div>
        </button>
      </div>
    </div>
    <div
      aria-label="loading-container"
      class="loading-container hide"
    >
      <div
        class="overlay"
      />
      <img
        aria-label="loading-logo"
        class="rotating-logo"
        src="/assets/images/iconhive.png"
      />
      <div
        class="loading-text"
      >
        Loading...
      </div>
      <div
        class="operations"
      />
    </div>
  </div>
</DocumentFragment>
`;

exports[`transfer tests:
 Must show Transfer dialog(empty memo): Request Transfer empty memo 1`] = `
<DocumentFragment>
  <div
    aria-label="dialog-operation-component"
    class="operation"
  >
    <div>
      <h2
        aria-label="dialog-header-operation"
        class="dialog-header"
      >
        Transfer
      </h2>
    </div>
    <div
      class="operation_body"
    >
      <h3>
        Account
      </h3>
      <div
        aria-label="operation_item_content"
        class="operation_item_content "
      >
        @keychain.tests
      </div>
      <h3>
        To
      </h3>
      <div
        aria-label="operation_item_content"
        class="operation_item_content "
      >
        @theghost1980
      </div>
      <h3>
        Amount
      </h3>
      <div
        aria-label="operation_item_content"
        class="operation_item_content "
      >
        1.000 HIVE
      </div>
      <h3>
        Balance
      </h3>
      <div
        aria-label="operation_item_content"
        class="operation_item_content "
      >
        100 HIVE =&gt; 99.000 HIVE
      </div>
    </div>
    <div
      class="operation_footer no-whitelist"
    >
      <div
        aria-label="whitelist-operation"
        class="whitelist_operation"
      />
      <div
        class="operation_buttons "
      >
        <button
          aria-label="dialog_cancel-button"
          class="submit-button stroked "
        >
          <div
            class="button-label"
          >
            Cancel 
          </div>
        </button>
        <button
          aria-label="dialog_confirm-button"
          class="submit-button raised "
        >
          <div
            class="button-label"
          >
            Confirm 
          </div>
        </button>
      </div>
    </div>
    <div
      aria-label="loading-container"
      class="loading-container hide"
    >
      <div
        class="overlay"
      />
      <img
        aria-label="loading-logo"
        class="rotating-logo"
        src="/assets/images/iconhive.png"
      />
      <div
        class="loading-text"
      >
        Loading...
      </div>
      <div
        class="operations"
      />
    </div>
  </div>
</DocumentFragment>
`;

exports[`transfer tests:
 Must show Transfer dialog(regular memo): Request Transfer regular memo 1`] = `
<DocumentFragment>
  <div
    aria-label="dialog-operation-component"
    class="operation"
  >
    <div>
      <h2
        aria-label="dialog-header-operation"
        class="dialog-header"
      >
        Transfer
      </h2>
    </div>
    <div
      class="operation_body"
    >
      <h3>
        Account
      </h3>
      <div
        aria-label="operation_item_content"
        class="operation_item_content "
      >
        @keychain.tests
      </div>
      <h3>
        To
      </h3>
      <div
        aria-label="operation_item_content"
        class="operation_item_content "
      >
        @theghost1980
      </div>
      <h3>
        Amount
      </h3>
      <div
        aria-label="operation_item_content"
        class="operation_item_content "
      >
        1.000 HIVE
      </div>
      <h3>
        Balance
      </h3>
      <div
        aria-label="operation_item_content"
        class="operation_item_content "
      >
        100 HIVE =&gt; 99.000 HIVE
      </div>
      <h3>
        Memo
      </h3>
      <div
        aria-label="operation_item_content"
        class="operation_item_content "
      >
        Regular memo!
      </div>
    </div>
    <div
      class="operation_footer no-whitelist"
    >
      <div
        aria-label="whitelist-operation"
        class="whitelist_operation"
      />
      <div
        class="operation_buttons "
      >
        <button
          aria-label="dialog_cancel-button"
          class="submit-button stroked "
        >
          <div
            class="button-label"
          >
            Cancel 
          </div>
        </button>
        <button
          aria-label="dialog_confirm-button"
          class="submit-button raised "
        >
          <div
            class="button-label"
          >
            Confirm 
          </div>
        </button>
      </div>
    </div>
    <div
      aria-label="loading-container"
      class="loading-container hide"
    >
      <div
        class="overlay"
      />
      <img
        aria-label="loading-logo"
        class="rotating-logo"
        src="/assets/images/iconhive.png"
      />
      <div
        class="loading-text"
      >
        Loading...
      </div>
      <div
        class="operations"
      />
    </div>
  </div>
</DocumentFragment>
`;
